Add composite trip

sebastianbarry opened the pull request:
This PR addresses the issues started in https://github.com/shankari/e-mission-server/pull/3#issuecomment-1434142457 - that PR is good to be considered old, and this new PR is addressing 2 topics:
1. Adding `confirmed_place`
2. Adding `composite_trip`

With this, the `CREATE_PLACE_OBJECTS` pipeline stage has been removed and replaced with a new pipeline stage `CREATE_COMPOSITE_TRIPS`. Jack and I decided that because ["since "composite" is representing a new class of object for the presentation layer, it warranted a new pipeline stage"](https://github.com/shankari/e-mission-server/pull/3#issuecomment-1434142457)

With this new pipeline step, there is a new matcher.py function: `create_composite_objects` which takes the already created confirmed_trip objects and appends the destination's confirmed_place object to it

sebastianbarry committed:
Create confirmedplace.py

To create confirmedplace.py, I also created expectedplace.py and inferredplace.py

sebastianbarry committed:
Added confirmedplace in entry.py and builtin_timeseries.py

sebastianbarry committed:
Removed expectedplace.py and inferredplace.py

Because these files are not needed as places do not need inferred and expected, i am removing them

sebastianbarry committed:
Added confirmedplace into the pipeline

pipelinestate.py
- Added a 17th ENUM to the PipelineStages named "CREATE_PLACE_OBJECTS"

reset.py
- Added a Pipeline stage to the stages_with_fuzz array for CREATE_PLACE_OBJECTS

common.py
- Added a "create_place_objects" function after all the other pipeline stage functions

intake_stage.py
- Added a stage for the intake to store the pipeline time for CREATE_PLACE_OBJECTS, and call the create_place_object function

confirmedplace.py
- Removed expected_place and primary_section because @shankari  said in our last meeting that the confirmedplace object does not need these

matcher.py
- Created the create_place_objects function which follows the same format as create_confirmed_objects, but only for trips. It uses the same existing mark_confirmed_object_creation_done/failed functions, and the same get_time_range_for_confirmed_object_creation
- Added a create_confirmed_places function which follows the same format as create_confirmed_trips; I removed any data which does not exist in the confirmed_places emission/core/wrapper/confirmedplace.py data model

sebastianbarry committed:
First draft

JGreenlee committed:
create confirmed_places during CREATE_CONFIRMED_OBJECTS

-- additions matching function genericized for both trips and places

JGreenlee committed:
implement CREATE_COMPOSITE_OBJECTS

- this new stage in the pipeline will create composite trip objects, after confirmed objects have been created

sebastianbarry commented:
One thing I noticed:

The data model on the server does not quite match the model being received on the phone
| Server | Phone |
:---:|:---:
| <img width="764" alt="image" src="https://user-images.githubusercontent.com/61334340/219702427-7068edf4-23a6-4688-ae48-beb3efcad62f.png"> | <img width="479" alt="image" src="https://user-images.githubusercontent.com/61334340/219702661-5ec48d32-1dda-436e-b911-365106f7baf9.png"> |

That being said, I believe we can remove 
```python
                  "user_input": ecwb.WrapperBase.Access.WORM,
                  "additions": ecwb.WrapperBase.Access.WORM
```
from the model per https://github.com/shankari/e-mission-server/pull/3#pullrequestreview-1300217599
> note that we are not supporting place-level inferences right now, so no need to include these either.

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -63,32 +63,88 @@ def handle_single_most_recent_match(confirmed_trip, ui):
 
 def handle_multi_non_deleted_match(confirmed_trip, ui):
     logging.debug(f"handling user input {ui} for {confirmed_trip}")
```
Nothing in this is trip specific, so maybe we should rename the variable to `confirmed_object` or `confirmed_trip_or_place` for consistency

JGreenlee replied:
Agreed.
I've been using "timeline entry" as a generic term for either a trip or place object

sebastianbarry replied:
Resolved in https://github.com/shankari/e-mission-server/pull/4/commits/5cdbdeb219c85714684677ea9f91f083397141be

I went with `confirmed_object`, but I am also good to change it to `timeline_entry` if we feel that is a better name

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -63,32 +63,88 @@ def handle_single_most_recent_match(confirmed_trip, ui):
+        logging.debug("len(confirmedTrips) == 0, early return")
+        return None
+    for ct in confirmedTrips:
+        logging.info("End place type for trip is %s" %  type(ct.data.end_place))
+        end_place = esda.get_entry(esda.CONFIRMED_PLACE_KEY, ct["data"]["confirmed_place"])
+        composite_trip_dict = copy.copy(ct)
+        del composite_trip_dict["_id"]
+        composite_trip_dict["metadata"]["key"] = "analysis/composite_trip"
+        composite_trip_dict["data"]["confirmed_place"] = end_place
+
+        # later we will want to put section & modes in composite_trip as well
+
+        composite_trip_entry = ecwe.Entry(composite_trip_dict)
+        # save the entry
+        ts.insert(composite_trip_entry)
```
This is the correct high-level approach. Note, however, that eventually, we do want to put the trajectory in here. But we can defer that until we are done with the current sprint for place and it is tested end to end.

sebastianbarry replied:
Added to to-do section

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -63,32 +63,88 @@ def handle_single_most_recent_match(confirmed_trip, ui):
     try:
-        last_expected_trip_done = create_confirmed_trips(user_id, time_query)
-        if last_expected_trip_done is None:
+        # we will use the same query for trips and places, but using 'exit_ts' instead 'end_ts'
+        time_query.timeType = "data.exit_ts"
+        last_expected_place_done = create_confirmed_places(user_id, time_query)
+        time_query.timeType = "data.end_ts"
+        last_expected_trip_done = create_confirmed_trips(user_id, time_query, last_expected_place_done["_id"])
+        if last_expected_trip_done is None or last_expected_place_done is None:
```
This is correct from a high level perspective, but there are a couple of minor wrinkles that we (1) need to be careful about, (2) see how we have done before, and (3) test.

1. the last place (aka the place we are in right now) will not have an exit timestamp. So I would expect that it would not be returned by this query. That implies that we will not create a confirmed place for it. So if the user added any time use entries to the last place, I would expect that they would not visible to the user after the timeline has finished processing.
    a. Trip ends at 5:30, will not leave until the next morning
    b. Pipeline runs, we create a cleaned place with enter = 5:30
    c. user enters a time use entry
    d. entry is pushed to server, pipeline runs 8:30pm
    e. when we process the entry, it doesn't match any confirmed place because the cleaned place will not match the exit_ts query
    f. next morning we take the new trip
    g. next time pipeline runs, we will create a confirmed trip, and it will match the existing user input
The part that I am worried about is between steps d and f because the pipeline may think that it had finished processing the entry, so it will not be returned as an unprocessed entry and it will not be matched either, so it may seem to "disappear". We need to test this super carefully.
2. for the last trip, when we try to find the composite place at `end_place = esda.get_entry(esda.CONFIRMED_PLACE_KEY, ct["data"]["confirmed_place"])`, we won't find anything. So I think that we will not even display the final place?
3. should we mark the `done` timestamp as the trip or the place? If we use the trip, this code works except for the issue above because we will have generated confirmed trips until the final trip and the confirmed place will be generated after the next trip is done. But if we want to address the issue above, we may want to mark the `done` with the `enter_ts` of the last place instead. We should check on how we have done this for other stages that mix trips and places, namely `CLEAN_AND_RESAMPLE`

JGreenlee replied:
So the last `cleaned_place` (where we are right now) will have an `enter_ts` but no `exit_ts`. I thought that we wouldn't want to create a `confirmed_place` that is incomplete in this way.

But if we do want to, can't we just query for cleaned places using  `enter_ts` instead? The last trip's `end_ts` is always the same as the last place's `enter_ts` right?

Then we would have a trailing `confirmed_place`, and it wouldn't have an `exit_ts`. The input matching functions could accommodate for that by just skipping end checks if `exit_ts` doesn't exist

shankari replied:
I think that the two options that would work are:
1) do what we do for `CLEAN_AND_RESAMPLE`, which is that we keep the trip timestamp as the primary BUT when we handle a trip, we also handle it succeeding place in the step. So basically have a `create_confirmed_trip` and within that function, while processing trips also create the confirmed place that follows it.
2) or use `enter_ts` for places and keep it seperate

Advantage of (1) that you can look at `CLEAN_AND_RESAMPLE` for a template and we know it works
Advantage of (2) is that it is cleaner conceptually since we are working on parts of the timeline independently and this does not appear to have cross component correlations as we do in the `CLEAN_AND_RESAMPLE`

shankari commented on emission/core/wrapper/compositetrip.py:
```diff
@@ -0,0 +1,31 @@
+    props.update({#      # confirmedplace stuff
+                  "confirmed_place": ecwb.WrapperBase.Access.WORM, # object contains all properties for the destination confirmed_place object
+                  #      # sections stuff
+                  "cleaned_section": ecwb.WrapperBase.Access.WORM,
+                  "inferred_section": ecwb.WrapperBase.WORM,
+                  "inferred_mode": ecwb.WrapperBase.WORM, # inferred by mode inference algo
+                  "confirmed_mode": ecwb.WrapperBase.WORM, # confirmed by user
```
we do want to have the trajectory here as well in the future

sebastianbarry replied:
Let's make this a to-do item

shankari commented on emission/core/wrapper/confirmedplace.py:
```diff
@@ -0,0 +1,28 @@
+    props = ecwp.Place.props
+    props.update({"raw_place": ecwb.WrapperBase.Access.WORM,
+                  "cleaned_place": ecwb.WrapperBase.Access.WORM,
+                  "inferred_labels": ecwb.WrapperBase.Access.WORM,
+                  "expectation": ecwb.WrapperBase.Access.WORM,
```
as we discussed, we are not planning to run label inferences on place right now since we don't have a use case for it, so we can remove these.

shankari replied:
please remove the others below as well

sebastianbarry replied:
Removed these in https://github.com/shankari/e-mission-server/pull/4/commits/64f1f5153ac795255174f39240cfddb7568a7bd2

I removed the following properties in `confirmed_place`
- inferred_labels
- expectation
- inferred_primary_mode
- user_input

Because we don't need inferred labels and expectation, and we also don't need a primary mode at a place, or a 1:1 user input at a place either

I did leave `additions` because we want this for the time-use and missing-trip surveys

shankari commented on emission/storage/timeseries/builtin_timeseries.py:
```diff
@@ -369,7 +371,7 @@ def insert(self, entry):
         Inserts the specified entry and returns the object ID 
         """
-        logging.debug("insert called with entry of type %s" % type(entry))
+        logging.info("insert called with entry of type %s" % type(entry))
```
please revert, no need to have log spew at the `INFO` log level

JGreenlee replied:
Understood. I was just using this to confirm it was working at the time

Is there a flag or option we can set to temporarily show debug logs, so we don't have to use `.info`?

sebastianbarry replied:
Resolved in https://github.com/shankari/e-mission-server/pull/4/commits/3fbc72d73cd4d0e73f5ceb336e8c151e2da1ebc9

shankari replied:
@JGreenlee the logging in general is controlled by files in `conf/log`. What you see on the console is not the full extent of the logging.

shankari commented on emission/storage/timeseries/builtin_timeseries.py:
```diff
@@ -378,9 +380,9 @@ def insert(self, entry):
 		(entry, entry["user_id"], self.user_id))
         else:
-            logging.debug("entry was fine, no need to fix it")
+            logging.info("entry was fine, no need to fix it")
```
Ditto

sebastianbarry replied:
Resolved in https://github.com/shankari/e-mission-server/pull/4/commits/3fbc72d73cd4d0e73f5ceb336e8c151e2da1ebc9

shankari commented on emission/storage/timeseries/builtin_timeseries.py:
```diff
@@ -378,9 +380,9 @@ def insert(self, entry):
+            logging.info("entry was fine, no need to fix it")
 
-        logging.debug("Inserting entry %s into timeseries" % entry)
+        logging.info("Inserting entry %s into timeseries" % entry)
```
Ditto

sebastianbarry replied:
Resolved in https://github.com/shankari/e-mission-server/pull/4/commits/3fbc72d73cd4d0e73f5ceb336e8c151e2da1ebc9

shankari reviewed (changes requested):
Everything else looks good to me. The high level structure of adding a new pipeline stage is correct. I just want to be careful about the timestamp used to mark the stage as done (see my giant note). Note that before this, we have not actually displayed places, so even if the cleaned place was created later, it might be OK, but now we are displaying the places, so we do need to make sure that the final confirmed place is created in the pipeline stage.

sebastianbarry committed:
Change confirmed_trip variable to confirmed_object for consistency

Per https://github.com/shankari/e-mission-server/pull/4#discussion_r1110032259 we will rename this to confirmed_object since it does not specifically need to be a trip

sebastianbarry committed:
Remove unnecessary confirmed_place properties

Per https://github.com/shankari/e-mission-server/pull/4#discussion_r1110057488 we will remove inferred_labels, expectation, inferred_primary_mode, and user_input since confirmed_place will not need that for now

sebastianbarry committed:
Remove info log spew by replacing Logging.info changes back to Logging.debug

Per https://github.com/shankari/e-mission-server/pull/4#discussion_r1110061541 we will remove Logging.info and replace with Logging.debug statements

JGreenlee commented on emission/core/wrapper/confirmedplace.py:
```diff
@@ -12,15 +12,7 @@ class Confirmedplace(ecwp.Place):
-                  "inferred_primary_mode": ecwb.WrapperBase.Access.WORM,
-# the user input will have all `manual/*` entries
-# let's make that be somewhat flexible instead of hardcoding into the data model
-                  "user_input": ecwb.WrapperBase.Access.WORM,
```
We should keep `user_input`. We don't have a use case yet but I believe we said we are going to support 1:1 inputs on places, as well as the 1:n 'additions'
I also thought we didn't need `confidence_threshold`.

I have revised in d016b3db1e38fba3b6ee2c4f172a8d0ca2952e85

sebastianbarry commented:
# To-do:

* Add `trajectory` (or an object representing the trajectory) as a property of `emission/core/wrapper/compositetrip.py`

JGreenlee commented on emission/core/wrapper/confirmedplace.py:
```diff
@@ -12,15 +12,7 @@ class Confirmedplace(ecwp.Place):
                   "cleaned_place": ecwb.WrapperBase.Access.WORM,
-                  "inferred_labels": ecwb.WrapperBase.Access.WORM,
-                  "expectation": ecwb.WrapperBase.Access.WORM,
                   "confidence_threshold": ecwb.WrapperBase.Access.WORM,
```
I thought we wouldn't need `confidence_threshold` either?

JGreenlee committed:
update confirmed_place properties
- we will keep user_input
- 'cleaned place' will be used since there is no 'expected place'
- clarify comments/ log statements

JGreenlee commented:
#### One important issue that needs to be ironed out at some point:
Every time the pipeline runs now, we will create `n` composite trips for `n` confirmed trips. We don't remove previous entries, so they accumulate over time and make duplicates. We obviously don't want that to happen but for the first draft I didn't look into the issue too far - I'm guessing there is some preferred solution to remove the previous entries when new ones are being added.

JGreenlee committed:
add a `locations` property to composite trip

JGreenlee committed:
correct confirmed_place generation

JGreenlee commented on emission/analysis/userinput/matcher.py:
```diff
@@ -61,34 +62,105 @@ def handle_single_most_recent_match(confirmed_trip, ui):
+        composite_trip_entry = ecwe.Entry(composite_trip_dict)
+        # save the entry
+        ts.insert(composite_trip_entry)
+    epq.mark_composite_object_creation_done(user_id, None)
```
When marking the composite creation stage done, we should pass the timestamp of the last place instead of "None".

I believe this will fix the duplication issues

JGreenlee committed:
use enter_ts to query places and mark stage done

JGreenlee committed:
revise input matching functions

JGreenlee committed:
add keys for place user input

JGreenlee committed:
revise input matching

JGreenlee committed:
modify user input tests

This is to make sure that our test cases only expect matches when the userinput starts before the end of the trip.
If the user input starts exactly when the trip ends, we will no longer match it to that trip (it will match to the trip's end place)
https://github.com/e-mission/e-mission-phone/pull/919#issuecomment-1441228756
https://github.com/e-mission/e-mission-phone/pull/919#issuecomment-1462942224

JGreenlee committed:
Merge branch 'add_trip_place_additions' into add-composite-trip

JGreenlee committed:
unify tripuserinput and placeuserinput

- since these are basically the same, they will both inherit from the generic Userinput

JGreenlee committed:
add logic to check places for inputs

JGreenlee committed:
revise unit tests to work with places matching

JGreenlee committed:
fix matching on last place

JGreenlee committed:
expand tests to check trip&place additions

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -38,57 +39,132 @@ def match_incoming_inputs(user_id, timerange):
         # TODO: Decide whether to error or to warn here
         logging.warn("Invalid status found in user input %s, moving forward anyway" % ui)
 
+def create_composite_objects(user_id):
+    time_query = epq.get_time_range_for_composite_object_creation(user_id)
+    ts = esta.TimeSeries.get_time_series(user_id)
```
high level cleanup: I am not sure that this function belongs here. Composite trip is not specifically related to input matching. This seems like the `OUTPUT_GEN` stage to me, so I would put this into `emission/analysis/result` or `emission/analysis/plotting`

JGreenlee replied:
Agreed, it's out of place. I just didn't know where to put it 😅

`emission/analysis/plotting` looks appropriate since composite trip represents what is being presented to the user

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -38,57 +39,132 @@ def match_incoming_inputs(user_id, timerange):
+        # retrieve locations for the trajectory of the trip
+        time_query = estt.TimeQuery("data.ts", ct["data"]["start_ts"], ct["data"]["end_ts"])
+        locations = esda.get_entries(esda.CLEANED_LOCATION_KEY, user_id, time_query=time_query)
+        max_entries = 100; # we will downsample to 100 locations
```
later cleanup: you might want to make this be configurable. Not sure if it needs to be in the dynamic config, or can be in a file in `conf`

JGreenlee replied:
I'll try to figure out dynamic config server-side

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -38,57 +39,132 @@ def match_incoming_inputs(user_id, timerange):
+        time_query.timeType = "data.enter_ts"
+        processed_places = create_confirmed_places(user_id, time_query)
+        time_query.timeType = "data.end_ts"
+        last_expected_trip_done = create_confirmed_trips(user_id, time_query, processed_places)
         if last_expected_trip_done is None:
             logging.debug("after run, last_expected_trip_done == None, must be early return")
             epq.mark_confirmed_object_creation_done(user_id, None)
         else:
             epq.mark_confirmed_object_creation_done(user_id, last_expected_trip_done.data.end_ts)
```
So we are marking the stage as "done" with the timestamp of the last trip rather than the last place? Will we be showing places in the UI? Shouldn't the last processed be the last place timestamp then?

shankari replied:
Please add some comments clarifying this

JGreenlee replied:
The end_ts of the last trip is always the same as the enter_ts of the last place, right?
Whichever way we access it, that seems like the right value to use.
If we used the exit_ts of the last place, then on the next pipeline run with new trips, when we query trips for start_ts, we might leave off the next trip.

Is that all correct? If yes, I can add an explanatory comment.

shankari commented on emission/core/wrapper/compositetrip.py:
```diff
@@ -0,0 +1,32 @@
+                  #      # sections stuff
+                  "cleaned_section": ecwb.WrapperBase.Access.WORM,
+                  "inferred_section": ecwb.WrapperBase.Access.WORM,
+                  "inferred_mode": ecwb.WrapperBase.Access.WORM, # inferred by mode inference algo
```
typically, the mode is inferred at the section level, not the trip level, so that we can support multi-modal trips. So this should really be an array or a dict instead of a single field.

JGreenlee replied:
These fields don't do anything yet but @sebastianbarry and I were speculating about how it might look.
I am not familiar enough with it yet; I'd like to revisit it in context when we actually add this to the label screen

shankari commented on emission/core/wrapper/compositetrip.py:
```diff
@@ -0,0 +1,32 @@
+                  "inferred_section": ecwb.WrapperBase.Access.WORM,
+                  "inferred_mode": ecwb.WrapperBase.Access.WORM, # inferred by mode inference algo
+                  "confirmed_mode": ecwb.WrapperBase.Access.WORM, # confirmed by user
+# mode to be used for analysis; confirmed mode if we know factors for it, inferred mode otherwise
```
what `factors` do you mean? Also this cannot be the inferred mode because the inferred mode is an array/dict. It can potentially be the primary inferred mode (see https://github.com/e-mission/e-mission-docs/issues/476#issuecomment-738120752)

shankari commented on emission/core/wrapper/placeuserinput.py:
```diff
@@ -5,27 +5,23 @@
     geojson = []
     nullable = []
-    local_dates = ["enter_local_dt", "exit_local_dt"]
+    local_dates = ["start_local_dt", "end_local_dt"]
```
why is the placeuserinput using `start/end` instead of `enter/exit`?

JGreenlee replied:
To me, enter/exit doesn't make sense here (you don't enter and exit an input). It's also just easier to work with one set of terms for the purpose of input matching

But if it's important to keep that terminology, we'll revert this and refactor the input matching to account for it

shankari commented on emission/core/wrapper/userinput.py:
```diff
@@ -0,0 +1,31 @@
+             "label": ecwb.WrapperBase.Access.RO, # string summary of the survey
+             "version": ecwb.WrapperBase.Access.RO, # the survey version
+             "name": ecwb.WrapperBase.Access.RO, # the survey name
+             "xmlResponse": ecwb.WrapperBase.Access.RO, # the XML string representation of the survey response
+             "jsonDocResponse": ecwb.WrapperBase.Access.RO # the JSON representation of the survey response
```
For the record, I think we should eventually also have the user labels in a user input object instead of having separate labels. But I guess we can change this code then.

JGreenlee replied:
What do you mean by "user labels"? For multilabel?

shankari commented on emission/storage/decorations/trip_queries.py:
```diff
@@ -73,66 +76,87 @@ def get_stops_for_trip(key, user_id, trip_id):
-                user_input.data.end_ts, next_trip_obj.data.start_ts, end_checks))
+        next_entry_obj = _get_next_cleaned_timeline_entry(ts, tl_entry)
+        if next_entry_obj is not None:
+            if hasattr(next_entry_obj.data, 'end_ts'):
+                next_entry_end = next_entry_obj.data.end_ts
+            elif hasattr(next_entry_obj.data, 'exit_ts'):
```
isn't this repeated from earlier? Refactor?

JGreenlee replied:
I'll refactor into helper functions 'begin' and 'stop'

shankari commented on emission/storage/decorations/trip_queries.py:
```diff
@@ -146,7 +170,9 @@ def final_candidate(filter_fn, potential_candidates):
     # run
     sorted_pc = sorted(extra_filtered_potential_candidates, key=lambda c:c["metadata"]["write_ts"])
-    entry_detail = lambda c: c.data.label if "label" in c.data else c.data.start_fmt_time
+    entry_detail = lambda c: getattr(c.data, "label", \
+                                getattr(c.data, "start_fmt_time", \
+                                getattr(c.data, "enter_fmt_time", None)))
```
please add comment on why you use `getattr` (I assume to avoid AttributeNotFound exceptions)

JGreenlee replied:
Yes. If label doesn't exist, we check for a start or enter time

will address in cleanup

shankari commented on emission/storage/decorations/trip_queries.py:
```diff
@@ -162,17 +188,24 @@ def get_not_deleted_candidates(filter_fn, potential_candidates):
-    tq = estt.TimeQuery("data.start_ts", trip_obj.data.start_ts, trip_obj.data.end_ts)
+def get_user_input_for_timeline_entry_object(ts, timeline_entry, user_input_key):
+    # timeline_entry can be either a trip or place, so let's check for start/enter or end/exit
+    if (hasattr(timeline_entry.data, 'start_ts')):
+        start = timeline_entry.data.start_ts
+        end = timeline_entry.data.end_ts
+    else:
+        start = timeline_entry.data.enter_ts
+        end = timeline_entry.data.exit_ts
```
this code again seems to be duplicated

JGreenlee replied:
I'll refactor into helper functions 'begin' and 'stop'

shankari commented on emission/storage/decorations/trip_queries.py:
```diff
@@ -185,17 +218,25 @@ def get_user_input_from_cache_series(user_id, trip_obj, user_input_key):
-        return valid_user_input_for_trip(ts, trip_obj, user_input)
+def get_additions_for_timeline_entry_object(ts, timeline_entry):
+    # timeline_entry can be either a trip or place, so let's check for start/enter or end/exit
+    if (hasattr(timeline_entry.data, 'start_ts')):
+        start = timeline_entry.data.start_ts
+        end = timeline_entry.data.end_ts
+    else:
+        start = timeline_entry.data.enter_ts
+        exit_ts = timeline_entry.data.exit_ts
+        end = exit_ts if exit_ts is not None else EPOCH_MAXIMUM
```
Duplicated

JGreenlee replied:
I'll refactor into helper functions 'begin' and 'stop'

shankari commented on emission/storage/decorations/trip_queries.py:
```diff
@@ -185,17 +218,25 @@ def get_user_input_from_cache_series(user_id, trip_obj, user_input_key):
     return curried
 
-def get_trip_for_user_input_obj(ts, ui_obj):
+def get_confirmed_obj_for_user_input_obj(ts, ui_obj):
```
I thought we know what kind of user input it is (trip input or place input). It seems like it would be more efficient (and likely more correct) to pass that in, and not search through all trips and places.

JGreenlee replied:
Sure, I can optimize this. If we are going to differentiate trip and place inputs we might as well take advantage of it here

shankari commented on emission/tests/storageTests/TestTripQueries.py:
```diff
@@ -95,8 +95,8 @@ def testUserInputForTripOneInputFromCache(self):
         new_mc = ecul.Userlabel()
-        new_mc["start_ts"] = new_trip.data.start_ts + 1
-        new_mc["end_ts"] = new_trip.data.end_ts + 1
+        new_mc["start_ts"] = new_trip.data.start_ts + .25
+        new_mc["end_ts"] = new_trip.data.end_ts + .25
```
maybe add an inline comment (linked to the commit history) with the reason for this set of changes.

JGreenlee replied:
will address in cleanup

shankari reviewed (changes requested):
@JGreenlee @sebastianbarry I am merging this since it seems to be functionally correct.
However, I do have some high-level code structure comments that it would be good to address in a cleanup PR if possible.
